File: 0169-majority-element/0169-majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        int candidate = nums[0];
        int count = 1;

        // Find the majority element using Boyer-Moore Voting Algorithm
        for (int i = 1; i < nums.length; i++) {
            if (count == 0) {
                candidate = nums[i];
                count = 1;
            } else if (candidate == nums[i]) {
                count++;
            } else {
                count--;
            }
        }

        return candidate;

        // Arrays.sort(nums); // Sort the array
        
        // // The majority element will occupy the middle position
        // return nums[nums.length / 2];
    }
}

File: 0169-majority-element/0169-majority-element.py
class Solution:
#     def majorityElement(self, nums: List[int]) -> int:
#         # ele=0
#         # for x in nums:
#         #     c=0
#         #     for y in nums:
#         #         if x==y:
#         #             c+=1
#         #     if c>(len(nums)/2):
#         #         ele=x
#         #         break
#         # return ele
#         for x in nums:
#             count = 0
#             for y in nums:
#                 if x == y:
#                     count += 1
#             if count > len(nums) // 2:
#                 return x
    def majorityElement(self, nums: List[int]) -> int:
        count = 0
        candidate = None

        for num in nums:
            if count == 0:
                candidate = num
            count += (1 if num == candidate else -1)

        return candidate
